pkg/backend/clustermanagerconfiguration.go

```diff
@@ -0,0 +1,119 @@
+package backend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
+)
+
+type clusterManagerConfigurationBackend struct {
+	*backend
+
+	configIterator cosmosdb.ClusterManagerConfigurationDocumentIterator
+	jsonSerializer *kjson.Serializer
+}
+
+func newClusterManagerConfigurationBackend(b *backend) *clusterManagerConfigurationBackend {
+	return &clusterManagerConfigurationBackend{
+		backend:        b,
+		configIterator: nil, // created as needed
+		jsonSerializer: kjson.NewSerializerWithOptions(
+			kjson.DefaultMetaFactory,
+			scheme.Scheme,
+			scheme.Scheme,
+			kjson.SerializerOptions{}),
+	}
+}
+
+func (cmc *clusterManagerConfigurationBackend) try(ctx context.Context, backendDoc *api.BackendDocument) (bool, error) {
+	log := cmc.backend.baseLog
+
+	if backendDoc == nil {
+		cmc.configIterator = nil
+		return false, nil
+	}
+
+	if cmc.configIterator == nil {
+		var options *cosmosdb.Options
+		if backendDoc.ClusterManagerConfigurations != nil {
+			// Create a change feed iterator starting from where the last
+			// lease owner left off, as recorded in the Continuation value.
+			options = &cosmosdb.Options{
+				Continuation: backendDoc.ClusterManagerConfigurations.Continuation,
+			}
+		}
+		// Keep this around for as long as we retain the backend lease.
+		cmc.configIterator = cmc.dbClusterManagerConfigurations.ChangeFeed(options)
+	}
+
+	restConfig, err := cmc.backend.env.LiveConfig().HiveRestConfig(ctx, 1)
+	if err != nil {
+		log.Info(err) // TODO(hive): Update to fail once we have Hive everywhere in prod and dev
+		return false, nil
+	}
+
+	dh, err := dynamichelper.New(log, restConfig)
+	if err != nil {
+		return false, err
+	}
+
+	docs, err := cmc.configIterator.Next(ctx, -1)
+	if err != nil {
+		return false, err
+	}
+
+	workWasDone := false
+
+	if docs != nil {
```

Comment:
In this loop I see it takes the configuration and passes it directly to the AKS/Hive cluster. What is the plan for linking the configuration with RP clusterDocument? I can easily see AKS going bananas and the need to restore the state. We also need a way to know which configuration is linked to which cluster.

Pardon me, if I missed it, this PR is realy big.

Reply:
Yeah, it's subtle.  The assumption is the payload from OCM will include the correct Hive namespace for the target cluster.  All we have to do is forward it on.

So if we had to do disaster recovery, I expect we would first recreate the same Hive namespaces for existing clusters based on the RP `clusterdocument`s.  Then we just throw all the resources in the `ClusterManagerConfigurations` container at Hive and they would land in the correct namespaces.